src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,275 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <gsl/gsl_errno.h>
+#include <gsl/gsl_integration.h>
+#include <tuple>
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+/// Numerical integration algorithms
+namespace integrate {
+
+/// The integration result is returned in this format.
+struct Result {
+  double value;
+  double error;
+};
+
+namespace detail {
+template <typename... Args>
+class GslQuadAdaptiveImpl;
+
+template <typename Args, typename ArgsIndices =
+                             std::make_index_sequence<tmpl::size<Args>::value>>
+struct ExpandTupleImpl;
+
+template <typename... Args, size_t... ArgsIndices>
+struct ExpandTupleImpl<tmpl::list<Args...>,
+                       std::index_sequence<ArgsIndices...>> {
+  template <typename FunctionType>
+  static double apply(FunctionType&& function, const double x,
+                      const std::tuple<Args...>& args) {
+    return (*function)(x, std::get<ArgsIndices>(args)...);
+  }
+};
+
+template <typename... Args>
+double integrand(double x, void* params) {
+  auto args = reinterpret_cast<GslQuadAdaptiveImpl<Args...>*>(params);
+  return ExpandTupleImpl<tmpl::list<Args...>>::apply(args->function_, x,
+                                                     args->args_);
+}
+
+// The GSL functions require the integrand to have this particular function
+// signature. In particular, any extra parameters to the functions must be
+// passed as a void* and re-interpreted appropriately. We interpret the
+// pointer as an instance of GslQuadAdaptiveImpl and use it to retrieve the
+// integrand function and its extra parameters.
+// The class builds a numerical integrator with the maximum number of
+// subintervals passed into the constructor.
+template <typename... Args>
+class GslQuadAdaptiveImpl {
+ public:
+  GslQuadAdaptiveImpl(size_t max_intervals) noexcept
+      : max_intervals_(max_intervals) {
+    initialize();
+  }
+
+  GslQuadAdaptiveImpl() = delete;
+  GslQuadAdaptiveImpl(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl& operator=(const GslQuadAdaptiveImpl&) = default;
+  GslQuadAdaptiveImpl(GslQuadAdaptiveImpl&&) noexcept = default;
+  GslQuadAdaptiveImpl& operator=(GslQuadAdaptiveImpl&& rhs) noexcept = default;
+  ~GslQuadAdaptiveImpl() noexcept = default;
+
+  // clang-tidy: no runtime references
```

Comment:
This comment is to explain why the next line has a `// NOLINT` that silences a clang-tidy warning. Please add the `// NOLINT` to the next line :) Otherwise the clang-tidy check on GitHub will fail (it currently doesn't because this file is never compiled, see comment below).

Reply:
you can remove the `// clang-tidy` comment since you have the `NOLINT(...)`